File: Segment Tree Build II.cpp
/**
 * Definition of SegmentTreeNode:
 * class SegmentTreeNode {
 * public:
 *     int start, end, max;
 *     SegmentTreeNode *left, *right;
 *     SegmentTreeNode(int start, int end, int max) {
 *         this->start = start;
 *         this->end = end;
 *         this->max = max;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     *@param A: a list of integer
     *@return: The root of Segment Tree
     */
    SegmentTreeNode * build(vector<int>& A) {
        // write your code here
        if (A.empty())
        {
            return nullptr;
        }
        return doBuild(0, A.size() - 1, A);
    }
    SegmentTreeNode *doBuild(int start, int end, vector<int> &A)
    {
        if (start == end)
        {
            return new SegmentTreeNode(start, start, A[start]);
        }
        SegmentTreeNode *node = new SegmentTreeNode(start, end, 0);
        int mid = start + (end - start) / 2;
        node->left = doBuild(start, mid, A);
        node->right = doBuild(mid + 1, end, A);
        node->max = max(node->left->max, node->right->max);
        return node;
        
    }
};

File: Segment Tree Build II.py
"""
Definition of SegmentTreeNode:
class SegmentTreeNode:
    def __init__(self, start, end, max):
        self.start, self.end, self.max = start, end, max
        self.left, self.right = None, None
"""

class Solution:	
    # @oaram A: a list of integer
    # @return: The root of Segment Tree
    def build(self, A):
        # write your code here
        
        def doBuild(left, right):
            #if left > right:
            #    return None
            node = SegmentTreeNode(left, right)
            if left == right:
                node.max = A[left]
                return node
            else:
                mid = left + (right - left) / 2
                node.left = doBuild(left, mid)
                node.right = doBuild(mid + 1, right)
                node.max = max(node.left.max, node.right.max)
                return node
        if not A:
            return None
        return doBuild(0, len(A) - 1)